src/test/run-pass/never_transmute_never.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(never_type)]
+#![allow(dead_code)]
+#![allow(unreachable_code)]
+#![allow(unused_variables)]
+
+struct Foo;
+
+fn f(x: !) -> ! {
+    x
+}
+
+fn ub() {
+    // This is completely undefined behaviour,
+    // but we still want to make sure it compiles.
+    let x: ! = unsafe {
+        std::mem::transmute::<Foo, !>(Foo)
+    };
+    f(x)
+}
+
+fn main() {}
```

Comment:
This test doesn't seem to trigger the ICE on current nightly. Possibly because the unused functions get never codegen'd?

NB: I would have expected a compile-pass test instead of run-pass (and with suitable crate type to avoid the dummy main)

Reply:
Good point on both counts — I've updated the test.